### PR TITLE
docs(apisix): wrote Stream Routes in Key Concepts

### PR DESCRIPTION
**Pull request by kayx23**
* wrote Stream Routes in Key Concepts: https://api7-docs-git-docs-stream-routes-api7ai.vercel.app/apisix/key-concepts/stream-routes
* updated Routes (L7) in Key Concepts to compare to Stream Routes (L4): https://api7-docs-git-docs-stream-routes-api7ai.vercel.app/apisix/key-concepts/routes
* added doc placeholder for Protos in Key Concepts

### CODE REVIEW

File: apisix_versioned_docs/version-3.2.0/background-information/key-concepts/stream-routes.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+<div style={{textAlign: 'center'}}>
+<img src="https://static.apiseven.com/uploads/2023/04/28/NxlpM4Yf_stream_routes.svg" alt="APISIX and MySQL Server with one stream route restricting on client IP address" width="100%"/>
+</div>
+
+The `client_addr` field in the stream route filters on the client IP, ensuring that only requests originating from the specified IP address are allowed to pass through this route to the upstream MySQL server. It should be noted that while both clients have used the correct username (`johndoe`) and password (`johnpwd`) to connect to the MySQL server, only the client with the IP address that has been whitelisted is permitted to establish a connection.
+
+In addition to the `client_addr` field, there are other fields such as `server_addr` and `server_port` that can be configured to filter incoming requests before rejecting or forwarding them to upstream services. To learn more about these fields and other configuration options, please refer to the Admin API reference for Stream Route (coming soon).
```
Comment (navendu-pottekkat) [inline apisix_versioned_docs/version-3.2.0/background-information/key-concepts/stream-routes.md:22]: > please refer to the Admin API reference for Stream Route (coming soon).

Inconsistent capitalization? stream route vs Stream Route?

**Review by navendu-pottekkat — COMMENTED** [review]
Looks good. Left some comments.

File: apisix_versioned_docs/version-3.2.0/background-information/key-concepts/routes.md [hunk 2]
```diff
@@ -9,9 +9,9 @@ Explore additional resources at the end of the document for more information on
 
 ## Overview
 
-_Routes_ define paths to upstream services. In APISIX, routes are responsible for matching client requests based on configured rules, loading & executing the corresponding plugins, as well as forwarding requests to the specified upstream endpoints.
+_Routes_ define paths to upstream services. In APISIX, a route object is used to create a route when APISIX operates on the [application layer](https://en.wikipedia.org/wiki/Application_layer), as opposed to a [stream route](./stream-routes.md) object, which is used to create a route when APISIX functions as a stream proxy operating on the [transport layer](https://en.wikipedia.org/wiki/Transport_layer). Routes are responsible for matching client requests based on configured rules, loading & executing the corresponding plugins, as well as forwarding requests to the specified upstream endpoints.
```
[the remark this answers is not in the document]
Comment (kayx23) [reply apisix_versioned_docs/version-3.2.0/background-information/key-concepts/routes.md:12]: @navendu-pottekkat ampersand use in text is actually not recommended. 

Microsoft style guide: https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/a/ampersand
if i recall correctly, google recommends the same.

File: apisix_versioned_docs/version-3.2.0/background-information/key-concepts/stream-routes.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+In this document, you will learn the basic concept of stream routes in APISIX and why you may need them. 
+
+Explore additional resources at the end of the document for more information on related topics. 
+
+## Overview
+
+In APISIX, a stream route object is used to create a route when APISIX functions as a stream proxy, operating on the [transport layer](https://en.wikipedia.org/wiki/Transport_layer) for TCP and UDP connections. This is similar to the concept of [routes](./routes.md) when APISIX functions as an [application-layer](https://en.wikipedia.org/wiki/Application_layer) proxy.
+
+The following diagram illustrates the concept of a stream route object using an example of APISIX functioning as a stream proxy for a MySQL server. [MySQL](https://www.mysql.com) is a relational database that communicates with clients over the TCP protocol. For this example, APISIX is deployed at IP address `192.168.1.10` and configured to listen for TCP connections on port `9500`:
```
[the remark this answers is not in the document]
Comment (kayx23) [reply apisix_versioned_docs/version-3.2.0/background-information/key-concepts/stream-routes.md]: I've rephrase this paragraph a bit

File: apisix_versioned_docs/version-3.2.0/background-information/key-concepts/stream-routes.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+<div style={{textAlign: 'center'}}>
+<img src="https://static.apiseven.com/uploads/2023/04/28/NxlpM4Yf_stream_routes.svg" alt="APISIX and MySQL Server with one stream route restricting on client IP address" width="100%"/>
+</div>
+
+The `client_addr` field in the stream route filters on the client IP, ensuring that only requests originating from the specified IP address are allowed to pass through this route to the upstream MySQL server. It should be noted that while both clients have used the correct username (`johndoe`) and password (`johnpwd`) to connect to the MySQL server, only the client with the IP address that has been whitelisted is permitted to establish a connection.
```
[the remark this answers is not in the document]
Comment (kayx23) [reply apisix_versioned_docs/version-3.2.0/background-information/key-concepts/stream-routes.md:20]: @navendu-pottekkat these two are actually covering about the same talking point, so they were in the same paragraph.

File: apisix_versioned_docs/version-3.2.0/background-information/key-concepts/routes.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (navendu-pottekkat) [reply apisix_versioned_docs/version-3.2.0/background-information/key-concepts/routes.md:12]: Yes, it is not recommended. So I suggested replacing it with "and." 😅 

"As well as" can also be replaced with "and." Makes it less wordy.

File: apisix_versioned_docs/version-3.2.0/background-information/key-concepts/stream-routes.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (navendu-pottekkat) [reply apisix_versioned_docs/version-3.2.0/background-information/key-concepts/stream-routes.md:20]: Okay, if it does not make sense to split it, we can still make it less wordy:

> It should be noted that while both clients

to

> While both clients

File: apisix_versioned_docs/version-3.2.0/background-information/key-concepts/routes.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (kayx23) [reply apisix_versioned_docs/version-3.2.0/background-information/key-concepts/routes.md]: lol my bad, looked at the wrong block

File: apisix_versioned_docs/version-3.2.0/background-information/key-concepts/stream-routes.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (kayx23) [reply apisix_versioned_docs/version-3.2.0/background-information/key-concepts/stream-routes.md:20]:
```suggestion
The `client_addr` field in the stream route filters on the client IP, ensuring that only requests originating from the specified IP address are allowed to pass through this route to the upstream MySQL server. While both clients have used the correct username (`johndoe`) and password (`johnpwd`) to connect to the MySQL server, only the client with the IP address that has been whitelisted is permitted to establish a connection.
```

**Review by An-DJ — APPROVED** [review]

**Pull request merged**